<center>
<h1>Intelligent Data Analysis Project</h1>
<h4>Matej Bebej (50%), Marian Kurcina (50%)</h4>
</center>

## Table of Contents
Assignment    
Phase 1 - Exploratory Data Analysis
- 1.1 - Basic description of data and their characteristics
  - A - Data structure analysis 
  - B - Analysis of individual attributes
  - C - Paiwise data analysis
  - D - Potential predictors
  - E - Observations
- 1.2 - Identification of problems, data integration, and cleaning
  - A - Data problems
  - B - Data accuracy check
  - C - Outlier detection
- 1.3 - Formulation and statistical verification of hypotheses about the data
  - A - Hypotheses
      - Hypothesis 1 -
      - Hypothesis 2 -
  - B - Statistical power
      - Hypothesis 1 -
      - Hypothesis 2 - 
Conclusion
<br><br><br>

# Assignment

Oxygen saturation is a key indicator of the proper functioning of the respiratory and circulatory systems. When its value drops to a critically low level, it may indicate life-threatening conditions such as hypoxemia, respiratory failure, or severe infections. In such cases, immediate intervention is essential. Traditional monitoring is performed using pulse oximeters, which, however, can be affected by noise, motion artifacts, or may have limitations in certain clinical situations.

Modern machine learning–based approaches offer the possibility to estimate and predict critical oxygen saturation values with higher accuracy (critical oxygen saturation estimation). Models can utilize multimodal data, such as heart rate, respiratory rate, blood pressure, or sensor signals. By being trained on diverse datasets, it is possible to identify early warning signs of desaturation, filter out noise, and provide timely alerts even before oxygen saturation drops below a safe threshold.

The goal of this assignment is to become familiar with the issue of oxygen saturation monitoring, understand the contribution of artificial intelligence, and design a solution that could improve critical care and reduce risks associated with undiagnosed hypoxemia.

Each pair of students will work with an assigned dataset starting from Week 2. Your task is to predict the dependent variable “oximetry” (the predicted variable) using machine learning methods. In doing so, you will need to deal with various issues present in the data, such as inconsistent formats, missing values, outliers, and others.

The expected outcomes of the project are:

 - the best-performing machine learning model, and

 - a data pipeline for building it from the input data.

# Phase 1 – Exploratory Data Analysis

In [38]:
import pandas as pd
import numpy as np
import matplotlib as plt
observation = pd.read_csv("dataset/observation.csv", sep='\t')
patient = pd.read_csv("dataset/patient.csv", sep='\t')
station = pd.read_csv("dataset/station.csv", sep='\t')

residence                                current_location blood_group  job  \
0        NaN  (Decimal('-39.6140315'), Decimal('22.351713'))          A+  NaN   
1        NaN    (Decimal('69.634637'), Decimal('66.117558'))         AB-  NaN   
2        NaN   (Decimal('19.563271'), Decimal('-41.108627'))         AB-  NaN   
3        NaN   (Decimal('34.010482'), Decimal('129.946434'))          O-  NaN   
4        NaN   (Decimal('61.148803'), Decimal('115.371931'))          O+  NaN   

                          mail  user_id   birthdate               company  \
0          schachtkaren@web.de     1135  1923-01-03      Ruppert Stey GbR   
1             plopes@gmail.com      773  1918-05-23    Gonçalves e Filhos   
2       tanakayoko@hotmail.com     1532  2001-01-04              中村食品有限会社   
3        brandoncruz@yahoo.com     2025         NaN         Silva-Gardner   
4  serragliogiorgio@tiscali.it      638  1997-05-14  Ortese-Agnesi s.r.l.   

                     name          username               ssn  \
0         Raymund Benthin         hornigdan       226-93-4078   
1       Leonardo Mendonça     diasana-clara       35096471810   
2                   吉田 陽子          sayuri90       066-73-4209   
3            Scott Thomas           rossjoy       368-27-2375   
4  Dott. Maurilio Zanazzo  orlandogilardoni  SNSGNT39E71B896C   

           registration  station_ID  
0            2021-04-21         729  
1  12/16/2019, 00:00:00         668  
2            2023/01/30         482  
3           09 Feb 2024         284  
4            2022-03-09         572

## 1.1 - Basic description of data and their characteristics

### A - Data structure analysis

Analyze file structures and relationships (number, types, etc.), records (structure, number of records, number of attributes, types, etc.).

In [24]:
observation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12107 entries, 0 to 12106
Data columns (total 23 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   SpO₂                   12107 non-null  float64
 1   HR                     12107 non-null  float64
 2   PI                     12107 non-null  float64
 3   RR                     12107 non-null  float64
 4   EtCO₂                  12107 non-null  float64
 5   FiO₂                   12107 non-null  float64
 6   PRV                    12107 non-null  float64
 7   BP                     12107 non-null  float64
 8   Skin Temperature       12107 non-null  float64
 9   Motion/Activity index  12107 non-null  float64
 10  PVI                    12107 non-null  float64
 11  Hb level               12107 non-null  float64
 12  SV                     12107 non-null  float64
 13  CO                     12107 non-null  float64
 14  Blood Flow Index       12107 non-null  float64
 15  PP

From following output, we can deduce that the dataset "observation" includes:    
- 12107 entries (rows)    
- 23 atributes (columns)

The dataset contains following attributes for each record:

SpO₂, HR, PI, RR, EtCO₂, FiO₂, PRV, BP, Skin Temperature, Motion/Activity index, PVI, Hb level, SV, CO, Blood Flow Index, PVI, Hb level, SV, CO, Blood Flow Index, PPG waveform features, Signal Quality Index, Respiratory effort, O₂ extraction ratio, SNR, oximetry, latitude, longitude

There are no attributes with null values, all attributes are float type.

In [30]:
patient.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2128 entries, 0 to 2127
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   residence         0 non-null      float64
 1   current_location  2022 non-null   object 
 2   blood_group       2128 non-null   object 
 3   job               638 non-null    object 
 4   mail              2128 non-null   object 
 5   user_id           2128 non-null   int64  
 6   birthdate         1170 non-null   object 
 7   company           2128 non-null   object 
 8   name              2128 non-null   object 
 9   username          2128 non-null   object 
 10  ssn               2128 non-null   object 
 11  registration      2128 non-null   object 
 12  station_ID        2128 non-null   int64  
dtypes: float64(1), int64(2), object(10)
memory usage: 216.3+ KB


In [31]:
print(patient.isnull().sum())

residence           2128
current_location     106
blood_group            0
job                 1490
mail                   0
user_id                0
birthdate            958
company                0
name                   0
username               0
ssn                    0
registration           0
station_ID             0
dtype: int64


From following output, we can deduce that the dataset "patient" includes:    
- 2128 entries (rows)    
- 13 atributes (columns)

The dataset contains following attributes for records:

residence, current_location, blood_group, job, mail, user_id, birthdate, company, name, username, ssn, registration, station_ID

There are several attributes with null values:
- residence - this attribute is not needed since it contains no non-null values
- current_location - this attribute has 106 null values
- job - this attribute has 1490 null values
- birthday - this attribute has 958 null values

residence is float type.
current_location, blood_group, job, mail, birthdate, company, name, username, ssn, registration are object type.
user_id, station_ID are int type


In [32]:
station.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 772 entries, 0 to 771
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   location   772 non-null    object 
 1   code       770 non-null    object 
 2   revision   772 non-null    object 
 3   station    772 non-null    object 
 4   latitude   772 non-null    float64
 5   longitude  772 non-null    float64
dtypes: float64(2), object(4)
memory usage: 36.3+ KB


In [33]:
print(station.isnull().sum())

location     0
code         2
revision     0
station      0
latitude     0
longitude    0
dtype: int64


From following output, we can deduce that the dataset "station" includes:    
- 772 entries (rows)    
- 6 atributes (columns)

The dataset contains following attributes for records:

location, code, revision, station, latitude, longtitude

There is only one attribute which cntains null values:
- code - this attribute has 2 null values

location, code, revision, station is object type.
latitude, longtitude are float type.

### B - Analysis of individual attributes

For at least 10 selected significant attributes, analyze their distributions and basic descriptive statistics, and check whether they meet the required conditions and the expected range of measured values.

In [34]:
print(observation.describe())

               SpO₂            HR            PI            RR         EtCO₂  \
count  12107.000000  12107.000000  12107.000000  12107.000000  12107.000000   
mean      97.720877     78.139220      9.280387     16.420159     39.761272   
std        0.849157      4.782149      3.432338      1.287131      1.220140   
min       95.000000     60.000000      0.200000     12.000000     35.000000   
25%       97.073862     74.938806      6.543803     15.418921     38.840854   
50%       97.850262     78.124134      8.703433     16.493047     39.782823   
75%       98.359154     81.345487     12.159623     17.466124     40.708415   
max      100.000000    100.000000     20.000000     20.000000     45.000000   

               FiO₂           PRV            BP  Skin Temperature  \
count  12107.000000  12107.000000  12107.000000      12107.000000   
mean      61.836198    106.629738    106.142109         35.712456   
std       10.182272     24.520873      3.676340          0.606630   
min       21

In [35]:
print(patient.describe())

       residence      user_id   station_ID
count        0.0  2128.000000  2128.000000
mean         NaN  1074.037594   390.836936
std          NaN   611.516264   224.274660
min          NaN     0.000000     0.000000
25%          NaN   554.000000   193.000000
50%          NaN  1074.500000   394.000000
75%          NaN  1607.250000   587.250000
max          NaN  2127.000000   770.000000


In [36]:
print(station.describe())

         latitude   longitude
count  772.000000  772.000000
mean    28.652487    9.132671
std     23.967608   73.458813
min    -54.800000 -156.474320
25%     14.443685  -67.331220
50%     36.230840    8.874040
75%     46.372028   69.920535
max     65.933810  168.313660


Comment

### C - Pairwise data analysis

Identify relationships and dependencies between pairs of attributes.

In [ ]:
# code

Comment

### D - Potential predictors

Identify dependencies between the predicted variable and other variables (potential predictors).

### E - Observations

Comment

## 1.2 - Identification of problems, data integration, and cleaning

## A - Data problems

Identify and initially address data problems such as: unsuitable data structure, duplicate records that may arise after certain transformations, inconsistent formats, missing values, and outliers. Other (not listed) problems may also occur, or fewer problems than those mentioned.

In [ ]:
# code

Comment

## B -  Data accuracy check

Do the data contain abnormal values?
Do they include illogical relationships caused by data collection or annotation errors?

In [ ]:
# code

Comment

### C - Outlier detection

Try to solve the problem using at least two techniques, e.g.:
 - Removing outliers or extreme observations
 - Replacing outliers with boundary values of the distribution (e.g., 5%, 95%)

In [ ]:
# code

Comment

## 1.3 Formulation and statistical verification of hypotheses about the data

### A - Hypotheses

Formulate two hypotheses about the data in the context of the given prediction task. Verify the formulated hypotheses using appropriately chosen statistical tests.
Example formulation:
“FiO₂ has a higher average value in the state with oximetry than without it.”

#### Hypothesis 1

Text of the hypothesis

Test comment

In [ ]:
# code

Comment

#### Hypothesis 2

Text of the hypothesis

Test comment

In [ ]:
# code

Comment

### B - Statistical power

Verify whether your statistical tests have sufficient support from the data — that is, whether they have enough statistical power.

#### Hypothesis 1

Comment

In [5]:
# code

#### Hypothesis 2

Comment

In [6]:
# code

# Conclusion